File: lib/core-ops/bracket.ts
/**
 * Core Ops — Single-elimination bracket generator.
 *
 * Generates a balanced bracket tree for top-cut play.
 * Supports standard sizes: 4, 8, 16, 32 players.
 *
 * Seeding follows standard tournament bracket seeding:
 *   - Seed 1 is placed to potentially meet Seed 2 in the final
 *   - Seed 1 and Seed 4 are on the same side, Seed 2 and Seed 3 on the other
 */

// --- Types ---

export interface BracketMatchOutput {
  bracketRound: number;
  bracketPosition: number;
  player1Id: string | null;
  player2Id: string | null;
  feedsWinnerToRound: number | null;
  feedsWinnerToPosition: number | null;
}

// --- Internal ---

/**
 * Generate standard tournament bracket seeding order (0-indexed).
 *
 * For 8 players: [0, 7, 3, 4, 1, 6, 2, 5]
 * → Seed1 vs Seed8, Seed4 vs Seed5, Seed2 vs Seed7, Seed3 vs Seed6
 */
export function standardSeeding(numPlayers: number): number[] {
  if (numPlayers === 1) return [0];
  if (numPlayers === 2) return [0, 1];

  const half = numPlayers / 2;
  const upper = standardSeeding(half);
  const lower = standardSeeding(half);

  const result: number[] = [];
  for (let i = 0; i < half; i++) {
    result.push(upper[i]);
    result.push(numPlayers - 1 - lower[i]);
  }
  return result;
}

// --- Main ---

/**
 * Generate a single-elimination bracket from seeded players.
 *
 * @param seeds Ordered list of player IDs by seed (index 0 = seed 1)
 * @param topCutSize Number of players in top cut (must be power of 2)
 * @returns Array of BracketMatchOutput for all matches in the bracket
 */
export function generateSingleElimBracket(
  seeds: string[],
  topCutSize: number
): BracketMatchOutput[] {
  if (topCutSize < 2) {
    throw new Error(`top_cut_size must be at least 2, got ${topCutSize}`);
  }

  if (seeds.length < topCutSize) {
    throw new Error(
      `Need at least ${topCutSize} seeds for top cut of ${topCutSize}, got ${seeds.length}`
    );
  }

  // Find the next power of 2 for the bracket structure
  const bracketSize = Math.pow(2, Math.ceil(Math.log2(topCutSize)));
  const numRounds = Math.log2(bracketSize);
  
  // Pad seeds with nulls (byes) up to the bracket size
  const actualSeeds = seeds.slice(0, topCutSize);
  while (actualSeeds.length < bracketSize) {
    actualSeeds.push(null as any); // Use null for bye spots
  }
  
  const seedingOrder = standardSeeding(bracketSize);

  const matches: BracketMatchOutput[] = [];

  // Round 1: pair seeds according to bracket seeding
  const round1Matches = bracketSize / 2;
  for (let pos = 0; pos < round1Matches; pos++) {
    const seedAIdx = seedingOrder[pos * 2];
    const seedBIdx = seedingOrder[pos * 2 + 1];

    const player1 = actualSeeds[seedAIdx] || null;
    const player2 = actualSeeds[seedBIdx] || null;

    const nextRound = numRounds > 1 ? 2 : null;
    const nextPosition = nextRound ? Math.floor(pos / 2) : null;

    matches.push({
      bracketRound: 1,
      bracketPosition: pos,
      player1Id: player1,
      player2Id: player2,
      feedsWinnerToRound: nextRound,
      feedsWinnerToPosition: nextPosition,
    });
  }

  // Subsequent rounds: empty slots filled as matches complete
  for (let roundNum = 2; roundNum <= numRounds; roundNum++) {
    const matchesInRound = bracketSize / Math.pow(2, roundNum);
    for (let pos = 0; pos < matchesInRound; pos++) {
      const nextRound = roundNum < numRounds ? roundNum + 1 : null;
      const nextPosition = nextRound ? Math.floor(pos / 2) : null;

      matches.push({
        bracketRound: roundNum,
        bracketPosition: pos,
        player1Id: null,
        player2Id: null,
        feedsWinnerToRound: nextRound,
        feedsWinnerToPosition: nextPosition,
      });
    }
  }

  return matches;
}

File: core-ops-service/app/bracket.py
"""Single-elimination bracket generator.

Generates a balanced bracket tree for top-cut play.
Supports standard sizes: 4, 8, 16, 32 players.

Seeding follows standard tournament bracket seeding where:
  - Seed 1 is placed to potentially meet Seed 2 in the final
  - Seed 1 and Seed 4 are on the same side, Seed 2 and Seed 3 on the other
  - This pattern recursively applies for larger brackets
"""

from __future__ import annotations

import math

from .models import BracketMatchOutput, BracketRequest


def _standard_seeding(num_players: int) -> list[int]:
    """Generate standard tournament bracket seeding order.

    For a bracket of size N, returns a list of seed indices (0-based)
    arranged so that higher seeds meet lower seeds first.

    Example for 8 players: [0, 7, 3, 4, 1, 6, 2, 5]
    This means: Seed1 vs Seed8, Seed4 vs Seed5, Seed2 vs Seed7, Seed3 vs Seed6
    """
    if num_players == 1:
        return [0]
    if num_players == 2:
        return [0, 1]

    # Recursive: split the bracket into two halves
    half = num_players // 2
    upper = _standard_seeding(half)
    lower = _standard_seeding(half)

    result = []
    for i in range(half):
        result.append(upper[i])
        result.append(num_players - 1 - lower[i])

    return result


def generate_single_elim_bracket(request: BracketRequest) -> list[BracketMatchOutput]:
    """Generate a single-elimination bracket.

    Args:
        request: BracketRequest with seeds list and top_cut_size.

    Returns:
        List of BracketMatchOutput representing all matches in the bracket.

    Raises:
        ValueError: If top_cut_size is not a power of 2, or seeds don't match size.
    """
    size = request.top_cut_size

    # Validate power of 2
    if size < 2 or (size & (size - 1)) != 0:
        raise ValueError(f"top_cut_size must be a power of 2, got {size}")

    if len(request.seeds) < size:
        raise ValueError(
            f"Need at least {size} seeds for top cut of {size}, got {len(request.seeds)}"
        )

    num_rounds = int(math.log2(size))
    seeds = request.seeds[:size]  # Take only top N

    # Generate seeding order
    seeding_order = _standard_seeding(size)

    matches: list[BracketMatchOutput] = []

    # Round 1: pair seeds according to bracket seeding
    round1_matches = size // 2
    for pos in range(round1_matches):
        seed_a_idx = seeding_order[pos * 2]
        seed_b_idx = seeding_order[pos * 2 + 1]

        player1 = seeds[seed_a_idx] if seed_a_idx < len(seeds) else None
        player2 = seeds[seed_b_idx] if seed_b_idx < len(seeds) else None

        # Determine where the winner feeds into
        next_round = 2 if num_rounds > 1 else None
        next_position = pos // 2 if next_round else None

        matches.append(
            BracketMatchOutput(
                bracket_round=1,
                bracket_position=pos,
                player1_id=player1,
                player2_id=player2,
                feeds_winner_to_round=next_round,
                feeds_winner_to_position=next_position,
            )
        )

    # Subsequent rounds: empty slots that get filled as matches complete
    for round_num in range(2, num_rounds + 1):
        matches_in_round = size // (2**round_num)
        for pos in range(matches_in_round):
            next_round = round_num + 1 if round_num < num_rounds else None
            next_position = pos // 2 if next_round else None

            matches.append(
                BracketMatchOutput(
                    bracket_round=round_num,
                    bracket_position=pos,
                    player1_id=None,  # TBD — winner of previous round
                    player2_id=None,
                    feeds_winner_to_round=next_round,
                    feeds_winner_to_position=next_position,
                )
            )

    return matches
